src/bsoncxx/types/value.cpp

```diff
@@ -40,15 +45,32 @@ value::value(const value& rhs) {
 }
 
 value& value::operator=(const value& rhs) {
-    _type = rhs._type;
+    if (this == &rhs) {
+        return *this;
+    }
+
+    if (_type) {
```

Comment:
Can we factor this out (it is repeated below in the move assign operator) into a private method named `destroy` or similar? That method could also take care of restoring _type to nullopt.